Add forward AD inplace check and fix codegen

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59993 All remaining linear/element-wise formulas
* #60499 Fix out= variant forward grad detection
* **#60498 Add forward AD inplace check and fix codegen**

Differential Revision: [D29914593](https://our.internmc.facebook.com/intern/diff/D29914593)

albanD committed:
Add forward AD inplace check and fix codegen

[ghstack-poisoned]

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

albanD commented:
You can see codegened code [here](https://github.com/albanD/pytorch/commits/codegen_diff) which is the same commits but with the generated files checked in.

zou3519 commented:
PR body should contain more description about how the forward AD inplace check works and what fixes were applied. I think I know the gist from our discussions but a summary is nice and anyone who comes along in the future who finds this commit may be confused

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
+    def emit_orignal_self_definition() -> List[str]:
+        body: List[str] = []
+        if inplace:
+            body.append('c10::optional<at::Tensor> original_self;')
```
I remember Ed saying something a while ago about splitting off forward-mode AD into its own DispatchKey (to be separate from reverse-mode AD). Some of the optimizations we're doing here (clone only once if both forward mode and reverse mode need a close of the input) are difficult to do if we do split forward-mode AD into different dispatch keys. Is this a concern?

albanD replied:
We might have to give up such optimizations when we move to a different dispatch key indeed.
We might be able to unpack the SavedVariable in the Node we just created, but that would be more complex than the current code for sure.

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_orignal_self_definition() -> List[str]:
```
nit: original

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -304,6 +304,11 @@
 }
 """)
 
+FW_DERIVATIVE_BOOLEAN_TEMPLATE = CodeTemplate("""\
+auto ${boolean_var_name} = ${cond};
+(void)${boolean_var_name};
+""")
```
Is there a way to check that the compiler optimized away this code if it's not used?
```
auto _any_has_forward_grad_result = isFwGradDefined(self) || isFwGradDefined(other);
(void)_any_has_forward_grad_result;
```
I remember sometimes the compiler gets confused about if it's safe to optimize away things (e.g. if the compiler thinks isFwGradDefined has any side effects)

albanD replied:
From checking the assembly code, the calls to isFwGradDefined are not optimized out indeed.
I'll add the appropriate logic to avoid these calls for out= functions

zou3519 replied:
RIP, I wish the compiler were smarter

albanD replied:
I think it could be if they were inlined and only calling simple functions, but that's not the case here :/

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
+                if derivative.required_original_self_value:
+                    all_forward_grad_cond.append(get_any_has_forward_grad_name(derivative.var_name))
+
+            if all_forward_grad_cond:
+                body.append(f'if ({" || ".join(all_forward_grad_cond)}) {{')
+                body.append('  original_self = self.clone();')
+                body.append('}')
+
+        return body
+
```
We probably want a big note somewhere about what the structure of a (for example) VariableType::mul_ code looks like. Some things I had to remind myself about include: we need to clone `self` before the redispatch because otherwise self can get mutated

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -583,6 +588,23 @@ def emit_check_no_requires_grad(
+
+            if all_forward_grad_cond:
+                body.append(f'if ({" || ".join(all_forward_grad_cond)}) {{')
+                body.append('  original_self = self.clone();')
```
Is... the first argument always called self?

zou3519 replied:
It looks like it always is :O

albanD replied:
This is for inplace ops. And for these, it always is. And it is always the argument that is modified inplace.

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -618,9 +642,13 @@ def save_variables(
                 expr = f'{expr}.has_value() ? c10::optional<std::string>(std::string({expr}.value())) : c10::nullopt'
             guard = guard_for(arg)
             if guard is None:
+                if stmts_prepend:
+                    stmts.append(f'{stmts_prepend};')
                 stmts.append(f'grad_fn->{name} = {expr};')
             else:
                 stmts.append(f'if ({guard}) {{')
+                if stmts_prepend:
+                    stmts.append(f'  {stmts_prepend};')
```
I'm not familiar with this code -- what is an example of the `guard` here?

albanD replied:
These guards are used when there are multiple formulas for a given function so that we can avoid saving Tensors when that input doesn't require gradients.
For example, for addmm:
```c++
    if (grad_fn->should_compute_output(1)) {
      grad_fn->mat2_ = SavedVariable(mat2, false);
    }
    grad_fn->alpha = alpha;
    if (grad_fn->should_compute_output(2)) {
      grad_fn->mat1_ = SavedVariable(mat1, false);
    }
```

zou3519 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -759,12 +780,40 @@ def emit_fw_derivatives() -> List[str]:
                                        'take a single TensorList as input. All other cases are not supported right now.')
                 requires_fw_grad = "true"
+
+            content.append(FW_DERIVATIVE_BOOLEAN_TEMPLATE.substitute(
```
Nit: Could we inline FW_DERIVATIVE_BOOLEAN_TEMPLATE here? the snippet is pretty small and inlining would improve readability

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -67,6 +67,14 @@ class ForwardDerivative:
+    # This is only used by inplace operations
+    required_original_self_value: bool
+
+    # If this formula is specified by the user or if we are re-using the
```
nit: "specified by the user" -> "specified in derivatives.yaml"? A user of PyTorch cannot specify a formula, right?

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -67,6 +67,14 @@ class ForwardDerivative:
+
+    # If this formula is specified by the user or if we are re-using the
+    # out of place formula for inplace
+    is_exact_match: bool
```
Would prefer a better name on this. How about "reuses_outplace_formula"?

albanD replied:
Yes.
I re-used the one that was already in that function. But in this context it can be improved indeed.

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
-                # replace "result" from the formula by self
+                formula = fw_info.formula
+
+                def update_for_original(formula: str, postfix: str) -> str:
```
nit: I had to read the use sites to figure out what this is doing. How about calling this "replace_self_with_original_self"?

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
+                        formula = update_for_original(formula, "_p")
+                        formula = update_for_original(formula, "_t")
+
+                if re.search(IDENT_REGEX.format("original_self_p"), formula):
+                    required_original_self_value = True
+                else:
+                    required_original_self_value = False
```
nit: `required_original_self_value = re.search(IDENT_REGEX.format("original_self_p"), formula)`

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
-                    return f'{m.group(1)}self{m.group(2)}'
-                formula = re.sub(IDENT_REGEX.format("result"), repl, fw_info.formula)
+                    return f'{m.group(1)}self_p{m.group(2)}'
+                formula = re.sub(IDENT_REGEX.format("result"), repl, formula)
```
nit: repl -> replace_result_with_self_p

albanD replied:
Not sure, these small functions are just for the matching. I added a comment above explaining what the pair regex/function is doing. Is that good?

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -259,27 +267,53 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
+                formula = re.sub(IDENT_REGEX.format("result"), repl, formula)
+
+                required_primals = fw_info.required_inputs_primal
+                if re.search(IDENT_REGEX.format("self_p"), formula):
```
I'm confused: didn't we replace all instances of `self_p` with `original_self_p`?

albanD replied:
But then we changed all the "result" to "self_p".
Let me add a comment at the beginning explaining the overall logic for this function

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

albanD committed:
Update on "Add forward AD inplace check and fix codegen"




[ghstack-poisoned]

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -761,7 +762,7 @@
 - name: polygamma_(Tensor(a!) self, int n) -> Tensor(a!)
   self: grad * polygamma(n + 1, self)
-  result: auto_element_wise
+  result: self_t.mul_(polygamma(n + 1, original_self_p))
```
There used to be some `conj` here? But I'm guessing that's okay because polygamma doesn't support complex
It also looks like we avoid doing mul by self_t out of place and then copy_
Are these optimizations that we could automate with auto_element_wise/auto_linear as well?

albanD replied:
We have a grand total of 10 functions that have formulas for the inplace version specifically. So I don't think we want to bother adding such a logic to these automation.

soulitzer replied:
Ah, good point
What was is the reasoning behind having functions have formulas specifically for inplace anyway?

albanD replied:
A few of them are because there is no out-of-place version.
A few of them is for perf to avoid input cloning by providing an alternative implementation that only makes use for the output.

soulitzer reviewed (approved):
Looks good to me! Just have some small questions.

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -614,6 +614,23 @@ def emit_check_no_requires_grad(
             body.append(f'check_no_requires_grad({arg_name}, "{arg_name}", "{name}");')
         return body
 
+    def emit_original_self_definition() -> List[str]:
```
The clone is saved as a SavedVariable on the grad_fn. So it outlives that function.
But we do keep one extra refcount to that Tensor longer indeed. I don't think it matters here though as nothing will check the SavedVariable before a hook is used or the backward happens.

soulitzer replied:
Makes sense, I didn't realize that this would be considered `output=False` and lead to the `save_original=True` logic in SavedVariable

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -257,29 +265,71 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
             forward_derivatives = info.forward_derivatives
 
             if f.func.kind() == SchemaKind.inplace:
+                # For inplace functions there is a little bit of work to do:
```
The post_process function is called for every formula in the derivatives.yaml file. So at that time, we don't know yet what is the function we're currently writing the code for.
This function here is the first one where we know both the exact function we're writing code for (and if it is inplace) and which formula we're using for it (and if it is an exact match or not).

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29914593).

albanD committed:
Update on "Add forward AD inplace check and fix codegen"


Differential Revision: [D29914593](https://our.internmc.facebook.com/intern/diff/D29914593)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29914593).

zou3519 commented on tools/codegen/api/autograd.py:
```diff
@@ -257,29 +265,71 @@ def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]
 
             if f.func.kind() == SchemaKind.inplace:
+                # For inplace functions there is a little bit of work to do:
+                #  1) Validate the formula and make sure the input that is modified in not used:
+                #    - If there is a formula for the inplace variant of the function (is_exact_match == True) then
+                #      we make sure that the original value of the input that is being modified inplace (self_p) is
+                #      not used in the formula. Note that the formula can use "original_self_p" here and that would
+                #      trigger a clone of the original input.
+                #    - If we are re-using the out of place formula (is_exact_match == False) then we replace every
+                #      occurrence of self_p and self_t by original_self_p and original_self_t. These will be
+                #      populated by cloned version of the original input (either the clone done by the backward AD
+                #      logic if self is also used in a backward formula or a special clone that we add).
+                #  2) At this point, there cannot be a self_p in the formula.
+                #  3) Change "result" into "self_p" as by design, in the inplace function codegen, the result is
+                #     simply called self (as it is modified inplace).
+                #  4) Update the required primals data in case it used to contain "result" but should now contain
+                #     "self"
+                #  5) If it is not an exact match, the user formula is not modifying the existing forward grad
+                #     inplace as it should. So add some code that makes sure that we do so if the forward grad
+                #     already exists.
```
nit: an example here would help, it's hard to read through a chunk of text without being able to visualize what happens

zou3519 reviewed (commented):
Sorry for the delayed review -- only had a comment on the comment but otherwise this lgtm as well
